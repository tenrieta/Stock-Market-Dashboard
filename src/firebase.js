import { initializeApp } from "firebase/app";
import { getAuth } from "firebase/auth";
import { getFirestore, doc, setDoc, getDoc } from "firebase/firestore";

const firebaseConfig = {
  apiKey: process.env.REACT_APP_FIREBASE_API_KEY,
  authDomain: process.env.REACT_APP_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.REACT_APP_FIREBASE_PROJECT_ID,
  storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.REACT_APP_FIREBASE_APP_ID,
};

// Initialize Firebase
const firebaseApp = initializeApp(firebaseConfig);

export const auth = getAuth();
export const db = getFirestore();

export const saveFirestore = async (newData) => {
  try {
    await setDoc(doc(db, "users", auth.currentUser.uid), newData);
  } catch (e) {
    console.error("Error adding document: ", e);
  }
};

export const readFirestore = async () => {
  try {
    const docRef = doc(db, "users", auth.currentUser.uid);
    const docSnap = await getDoc(docRef);
    if (docSnap.exists()) {
      return docSnap.data();
    }
    return { error: "User not found..." };
  } catch (error) {
    return { error };
  }
};

export default firebaseApp;
