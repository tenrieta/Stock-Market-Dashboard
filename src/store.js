/* eslint-disable no-underscore-dangle */
import { applyMiddleware, compose, createStore } from "redux";
import thunk from "redux-thunk";
import rootReducer from "./reducers";

const store = createStore(
  rootReducer,
  compose(
    applyMiddleware(thunk), // Thunk is necessary for async action creators
    window.__REDUX_DEVTOOLS_EXTENSION__
      ? window.__REDUX_DEVTOOLS_EXTENSION__()
      : (f) => f // This is to make the Redux for Chrome extension work in an incognito window with disabled Redux extension
  )
);

export default store;
