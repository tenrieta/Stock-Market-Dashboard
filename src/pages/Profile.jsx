import React from "react";
import { useDispatch, useSelector } from "react-redux";

// Bootstrap
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Card from "react-bootstrap/Card";
import Form from "react-bootstrap/Form";
import InputGroup from "react-bootstrap/InputGroup";
import FormControl from "react-bootstrap/FormControl";
import Button from "react-bootstrap/Button";
import Alert from "react-bootstrap/Alert";

// Firebase
import { signOut } from "firebase/auth";
import { auth } from "../firebase";

// Redux
import { updateUser } from "../actions/userActions";

const Profile = function Profile() {
  // Get state for currentUser from Redux
  const currentUser = useSelector((state) => state.currentUser);
  // Get dispatch for updating the state
  const dispatch = useDispatch();

  // Local state for Form Data
  const [formData, setFormData] = React.useState({
    formLoading: false,
    formError: null,
    formSuccess: null,
    formValidated: false,
    email: currentUser.email || "",
    displayName: currentUser.displayName || "",
    photoURL: currentUser.photoURL || "",
    newPassword1: "",
    newPassword2: "",
    oldPassword: "",
  });

  // Change handler which updates local state
  // In order to make it universal for all fields, we dynamically use [fieldName]: fieldValue
  const handleChange = (event) => {
    const fieldName = event.target.name;
    const fieldValue = event.target.value;
    setFormData((prevState) => ({ ...prevState, [fieldName]: fieldValue }));
  };

  // Submit handler
  const handleSubmit = (event) => {
    // First preventing default form behavior
    event.preventDefault();

    // Validation with Bootstrap
    setFormData((prevState) => ({ ...prevState, formValidated: true }));
    if (event.currentTarget.checkValidity() === false) {
      event.stopPropagation();
      return null;
    }

    // Verification of passwords. Password is updated only when filled out.
    if (
      (formData.newPassword1 || formData.newPassword2) &&
      formData.newPassword1 !== formData.newPassword2
    ) {
      setFormData((prevState) => ({
        ...prevState,
        formError: "The passwords don't match.",
      }));
      return null;
    }

    // Loading state prevents accidental double clicks on the submit button
    setFormData((prevState) => ({ ...prevState, formLoading: true }));

    // We send the data to the Action Creator, which takes care of the async connection with Firebase and Firestore
    // The Action creator then dispatches the action to Redux
    dispatch(updateUser(formData, setFormData, currentUser.firestore));

    return null;
  };

  // Sign out
  // When Firebase signs the user out, nulled user data is dispatched to Redux from App.jsx, where we have a listener onAuthStateChanged
  const handleSignout = () => {
    signOut(auth);
  };

  return (
    <Container>
      <Row className="justify-content-md-center">
        <Col className="col-lg-6 col-md-10 mx-auto">
          <Card className="mb-4">
            <Card.Body>
              <h2 className="text-center mb-4" id="profile">
                Update Profile
              </h2>
              {formData.formError && (
                <Alert variant="danger">{formData.formError}</Alert>
              )}
              {formData.formSuccess && (
                <Alert variant="success">{formData.formSuccess}</Alert>
              )}
              <Form
                noValidate
                validated={formData.formValidated}
                onSubmit={handleSubmit}
              >
                <Form.Group className="mb-3">
                  <Form.Label>Email</Form.Label>
                  <Form.Control
                    type="email"
                    required
                    value={formData.email}
                    onChange={handleChange}
                    name="email"
                    id="email"
                  />
                </Form.Group>
                <Form.Group className="mb-3">
                  <Form.Label>Display Name</Form.Label>
                  <Form.Control
                    type="text"
                    required
                    value={formData.displayName}
                    onChange={handleChange}
                    name="displayName"
                    id="displayName"
                  />
                </Form.Group>
                <Form.Group className="mb-3">
                  <Form.Label>Photo URL</Form.Label>
                  <Form.Control
                    type="text"
                    value={formData.photoURL}
                    onChange={handleChange}
                    name="photoURL"
                    id="photoURL"
                  />
                </Form.Group>
                <Form.Group className="mb-3">
                  <Form.Label>Password</Form.Label>
                  <Form.Control
                    type="password"
                    value={formData.newPassword1}
                    onChange={handleChange}
                    name="newPassword1"
                    id="newPassword1"
                    placeholder="Leave blank to keep the same"
                  />
                </Form.Group>
                <Form.Group className="mb-3">
                  <Form.Label>Password Confirmation</Form.Label>
                  <Form.Control
                    type="password"
                    value={formData.newPassword2}
                    onChange={handleChange}
                    name="newPassword2"
                    id="newPassword2"
                    placeholder="Leave blank to keep the same"
                  />
                </Form.Group>
                <p className="text-center fs-4 mb-3">
                  Ready to update your profile?
                </p>
                <p className="text-center mb-3">
                  To save changes, please verify your old password.
                </p>
                <InputGroup className="mb-3">
                  <FormControl
                    placeholder="Your old password"
                    required
                    type="password"
                    name="oldPassword"
                    id="oldPassword"
                    value={formData.oldPassword}
                    onChange={handleChange}
                    role="textbox"
                    title="oldPassword"
                  />
                  <Button
                    disabled={
                      formData.formLoading || formData.oldPassword === ""
                    }
                    variant="primary"
                    id="submit-button"
                    type="submit"
                  >
                    Update your profile
                  </Button>
                </InputGroup>
              </Form>
              <div className="text-center mb-3">
                <div className="fs-4 mb-3">Ready to leave?</div>
                <Button
                  className=""
                  variant="outline-secondary"
                  type="button"
                  onClick={handleSignout}
                  id="sign-out"
                >
                  Sign out
                </Button>
              </div>
            </Card.Body>
          </Card>
        </Col>
      </Row>
    </Container>
  );
};
export default Profile;
