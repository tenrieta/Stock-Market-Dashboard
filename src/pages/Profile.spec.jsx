import React from "react";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import {
  signInWithEmailAndPassword,
  reauthenticateWithCredential,
  EmailAuthProvider,
  updateProfile,
} from "firebase/auth";
import { auth } from "../firebase";
import store from "../store";
import Profile from "./Profile";

const mockStore = configureStore([]);

describe("SNAPSHOT TESTS", () => {
  test("Profile snapshot renders", async () => {
    const mockStoreInstance = mockStore({
      currentUser: { email: "jest@example.com" },
    });
    const component = renderer.create(
      <Provider store={mockStoreInstance}>
        <Profile />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("UNIT TESTS", () => {
  test("Update profile button is disabled if no password is provided.", () => {
    const { getByRole } = render(
      <Provider store={store}>
        <Profile />
      </Provider>
    );
    const oldPassword = getByRole("textbox", { name: "oldPassword" });
    const submitButton = getByRole("button", { name: "Update your profile" });
    expect(submitButton).toHaveAttribute("disabled");
    fireEvent.change(oldPassword, { target: { value: "abc123" } });
    expect(submitButton).not.toHaveAttribute("disabled");
  });
});

describe("INTEGRATION TESTS", () => {
  test("Update profile should throw error on wrong credentials.", async () => {
    let error = "";
    const incorrectPassword = "xxx";
    const correctPassword = "abc123";
    const newDisplayName = Math.floor(Math.random() * 1000 + 1).toString();
    await signInWithEmailAndPassword(auth, "jest@example.com", correctPassword);
    expect(auth.currentUser).toBeTruthy();
    expect(auth.currentUser.email).toBe("jest@example.com");

    try {
      const credential = EmailAuthProvider.credential(
        auth.currentUser.email,
        incorrectPassword
      );
      await reauthenticateWithCredential(auth.currentUser, credential);
      await updateProfile(auth.currentUser, { displayName: newDisplayName });
    } catch (err) {
      error = err.toString();
    }
    expect(error).toMatch(/auth\/wrong-password/);
  });

  test("Update displayName should work.", async () => {
    const correctPassword = "abc123";
    const newDisplayName = Math.floor(Math.random() * 1000 + 1).toString();
    await signInWithEmailAndPassword(auth, "jest@example.com", correctPassword);
    expect(auth.currentUser).toBeTruthy();
    expect(auth.currentUser.email).toBe("jest@example.com");

    try {
      const credential = EmailAuthProvider.credential(
        auth.currentUser.email,
        correctPassword
      );
      await reauthenticateWithCredential(auth.currentUser, credential);
      await updateProfile(auth.currentUser, { displayName: newDisplayName });
    } catch (err) {
      console.log(err.toString());
    }
    expect(auth.currentUser.displayName).toBe(newDisplayName);
  });
});
