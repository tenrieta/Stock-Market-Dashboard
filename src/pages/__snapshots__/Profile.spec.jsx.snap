// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT TESTS Profile snapshot renders 1`] = `
<div
  className="container"
>
  <div
    className="justify-content-md-center row"
  >
    <div
      className="col-lg-6 col-md-10 mx-auto col"
    >
      <div
        className="mb-4 card"
      >
        <div
          className="card-body"
        >
          <h2
            className="text-center mb-4"
            id="profile"
          >
            Update Profile
          </h2>
          <form
            className=""
            noValidate={true}
            onSubmit={[Function]}
          >
            <div
              className="mb-3"
            >
              <label
                className="form-label"
              >
                Email
              </label>
              <input
                className="form-control"
                id="email"
                name="email"
                onChange={[Function]}
                required={true}
                type="email"
                value="jest@example.com"
              />
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
              >
                Display Name
              </label>
              <input
                className="form-control"
                id="displayName"
                name="displayName"
                onChange={[Function]}
                required={true}
                type="text"
                value=""
              />
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
              >
                Photo URL
              </label>
              <input
                className="form-control"
                id="photoURL"
                name="photoURL"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
              >
                Password
              </label>
              <input
                className="form-control"
                id="newPassword1"
                name="newPassword1"
                onChange={[Function]}
                placeholder="Leave blank to keep the same"
                type="password"
                value=""
              />
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
              >
                Password Confirmation
              </label>
              <input
                className="form-control"
                id="newPassword2"
                name="newPassword2"
                onChange={[Function]}
                placeholder="Leave blank to keep the same"
                type="password"
                value=""
              />
            </div>
            <p
              className="text-center fs-4 mb-3"
            >
              Ready to update your profile?
            </p>
            <p
              className="text-center mb-3"
            >
              To save changes, please verify your old password.
            </p>
            <div
              className="mb-3 input-group"
            >
              <input
                className="form-control"
                id="oldPassword"
                name="oldPassword"
                onChange={[Function]}
                placeholder="Your old password"
                required={true}
                role="textbox"
                title="oldPassword"
                type="password"
                value=""
              />
              <button
                className="btn btn-primary"
                disabled={true}
                id="submit-button"
                type="submit"
              >
                Update your profile
              </button>
            </div>
          </form>
          <div
            className="text-center mb-3"
          >
            <div
              className="fs-4 mb-3"
            >
              Ready to leave?
            </div>
            <button
              className="btn btn-outline-secondary"
              disabled={false}
              id="sign-out"
              onClick={[Function]}
              type="button"
            >
              Sign out
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
